### PR TITLE
Add implicit FTPS support

### PR DESCRIPTION
**Pull request by alxpy**
SSL support added.

https://github.com/aio-libs/aioftp/issues/37

### CODE REVIEW

**Comment by asvetlov** [comment]
Any chance to add a test?

**Comment by alxpy** [comment]
I thought about tests, but I don't know a good way for testing transport without ssl support at the server part. Can you advise me how to implement it?

**Comment by pohmelie** [comment]
@alxpy, thanks for pr!

This pr is ¼ of what should be implemented for saying «we support ftps». There is two modes: explicit and implicit. And two sides: client and server. You implement implicit client ftps. Also, I'm sceptical about ftps, since it is pretty rare. But, anyway, I see no reasons to not to merge.

@asvetlov 
I'm also want some good tests for this, but as mentioned: without server side implementation tests are not possible in meaning even if they will be implemented they will mock server. I'm not sure that this will be good tests.

@alxpy, is there a chance you will implement implicit server side mode, so we can write tests and have half ftps support?

I will wait a day before merge. Maybe someone have strong point to reject this pr or @alxpy take care of implicit server side support.

**Comment by asvetlov** [comment]
You are right, mocked tests are useless for this case

**Comment by alxpy** [comment]
@pohmelie thanks for the detailed answer!

I will try to implement implicit server side ftps.

**Comment by rsichny** [comment]
lgtm (server side is useles, imo)

File: aioftp/client.py [hunk 1]
```diff
@@ -491,6 +495,13 @@ class Client(BaseClient):
 
     :param encoding: encoding to use for convertion strings to bytes
     :type encoding: :py:class:`str`
+
+    :param ssl: if given and not false, a SSL/TLS transport is created
```
Comment (rsichny) [inline aioftp/client.py]: please add link to https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_connection

Comment (alxpy) [reply aioftp/client.py]: done

**Comment by alxpy** [comment]
Implicit server side FTPS  added.

**Comment by alxpy** [comment]
I can't generate certificate which can work with Python 3.7 =(
https://github.com/aio-libs/aioftp/pull/81/files#diff-572606b9b77c96a18b4bf3eefabc6b25R65
https://bugs.python.org/issue34440

**Comment by alxpy** [comment]
@webknjaz I tried, `trustme` certs doesn't work without `ssl.match_hostname` monkey patching even with Py3.6

**Comment by webknjaz** [comment]
Why? Did you set up the hostname correctly?

File: tests/common.py [hunk 2]
```diff
@@ -42,12 +53,23 @@ def run_in_loop(s_args, s_kwargs, c_args, c_kwargs):
                         client.close()
                     loop.close()
 
+            # SSL monkey patching for tests with self signed certificates.
+            # Works only with Python older than 3.7 version.
+            old_match_hostname = ssl.match_hostname
+            ssl.match_hostname = lambda cert, hostname: True
```
Comment (webknjaz) [inline tests/common.py]: Better use `mock.patch` for such things. It's safer.

File: tests/test-throttle.py [hunk 3]
```diff
@@ -546,15 +546,15 @@ class SlowPathIO(aioftp.PathIO):
 
     async def worker(fname):
 
-        client = aioftp.Client(loop=loop)
-        await client.connect("127.0.0.1", PORT)
-        await client.login()
-        await client.download(
+        _client = aioftp.Client(loop=loop, ssl=client.ssl)
```
Comment (webknjaz) [inline tests/test-throttle.py]: There's no point/value in underscoring bunch of vars.

File: certificates/Certificate.crt [hunk 4]
```diff
@@ -0,0 +1,32 @@
+-----BEGIN CERTIFICATE-----
```
Comment (webknjaz) [inline certificates/Certificate.crt]: Use `trustme` for certs, it's more robust approach.

File: aioftp/server.py [hunk 5]
```diff
@@ -799,7 +805,8 @@ def __init__(self,
                  read_speed_limit_per_connection=None,
                  write_speed_limit_per_connection=None,
                  data_ports=None,
-                 encoding="utf-8"):
+                 encoding="utf-8",
+                 ssl=None):
```
Comment (webknjaz) [inline aioftp/server.py]: Pro tip: keep closing bracket on a separate line + use trailing commas — this makes diffs nicer.

**Comment by webknjaz** [comment]
@alxpy 
```python
ca = trustme.CA()
ca.issue_server_cert(u'127.0.0.1')
```

**Comment by alxpy** [comment]
Yes, I use hostname correctly. And any tests work, but not all.
I got `ssl.CertificateError: hostname '::1' doesn't match '127.0.0.1'`

File: aioftp/server.py (hunk 5 above)
Comment (alxpy) [reply aioftp/server.py]: Personaly I use this code style, but now I use code style of this project...

File: certificates/Certificate.crt (hunk 4 above)
Comment (alxpy) [reply certificates/Certificate.crt]: no problem

**Comment by webknjaz** [comment]
oh, that's easy

File: aioftp/client.py [hunk 6]
```diff
@@ -32,10 +32,11 @@
 logger = logging.getLogger(__name__)
 
 
-async def open_connection(host, port, loop, create_connection):
+async def open_connection(host, port, loop, create_connection, ssl=None):
```
[the remark this answers is not in the document]
Comment (alxpy) [reply aioftp/client.py]: something like `ssl_context` was my first idea, but prefer save original name

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_connection
https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server

**Comment by webknjaz** [comment]
@alxpy
```python
ssl_context = ...
...
...
ca = trustme.CA()
ca.issue_server_cert(u'127.0.0.1', u'::1', u'localhost')  # list all incarnations of localhost
ca.configure_trust(ssl_context)
```
Ref: https://trustme.readthedocs.io/en/latest/#trustme.CA.issue_server_cert

File: aioftp/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alxpy) [reply aioftp/client.py]: and here `ssl` really can be boolean =)

File: tests/test-throttle.py (hunk 3 above)
Comment (alxpy) [reply tests/test-throttle.py]: so how would you like to call a variable?

Comment (webknjaz) [reply tests/test-throttle.py]: keep it as is, I didn't realize that it's an "extended boolean"

**Comment by alxpy** [comment]
`trustme` works fine, thx

**Comment by pohmelie** [comment]
So... @webknjaz @rsichny @asvetlov you think this pr is ready to merge? lgtm. I'm pretty confused, that implicit ftps was so easy to implement.
Hats off to @alxpy :tophat:

File: tests/test-throttle.py (hunk 3 above)
Comment (webknjaz) [reply tests/test-throttle.py]: Oh, that was not about ssl

File: tests/test-throttle.py [hunk 7]
```diff
@@ -293,15 +293,15 @@ class SlowPathIO(aioftp.PathIO):
...
```
Comment (webknjaz) [inline tests/test-throttle.py]: Undo underscoring of `client` var. It's not related, nor needed.

**Comment by webknjaz** [comment]
@pohmelie FTPS seems to be just plain old FTP on top of TLS, nothing more. So stick SSL between TCP and the app and you're done. Simple and elegant.
SFTP would be a different story, though, as it works on top of SSH.


@alxpy I insist that wholesale renaming of `client` to `_client` in tests is useless and not necessary. It enlarges the diff and brings no value to tests at all.

**Comment by pohmelie** [comment]
@webknjaz 
I thought at first about explicit ftps: go to secure connection on unsecure command. That sounds pretty hard. Since there must be a method to go down to unsecure from secure. Of course, implicit ftps was pretty straight, but I did not mind, that it is so easy to implement.

**Comment by alxpy** [comment]
>  I insist ...

@webknjaz you can try rename `_client` to `client` and understand the reason for this change.

**Comment by webknjaz** [comment]
@alxpy so tell me: what is that?

**Comment by pohmelie** [comment]
@alxpy Thank you for great contribution! :tada:

**Pull request merged**